### ISSUE TITLE
something wrong with SparseEmbedding?

### ISSUE THREAD

**Issue by charleshm** [apache/incubator-mxnet#9266]
I wrote a embedding version of fm, but got bad results. I checked my code several times,  didn't find any bug. anybody help check it ? 

`    """ builds factorization machine network with proper formulation:
    y = w_0 \sum(x_i w_i) + 0.5(\sum\sum<v_i,v_j>x_ix_j - \sum<v_iv_i>x_i^2)
    F : field_size
    K : factor_size
    """
    x = mx.symbol.Variable("data") # None * F
    # 交叉项权重: D * K
    v = mx.symbol.Variable("v", shape=(num_features, factor_size), stype='row_sparse',
                           init=init_config['v'], lr_mult=lr_mult_config['v'],
                           wd_mult=wd_mult_config['v'])
    # 一阶权重: D * 1
    w = mx.symbol.var('w', shape=(num_features, 1), stype='row_sparse',
                      init=init_config['w'], lr_mult=lr_mult_config['w'],
                      wd_mult=wd_mult_config['w'])
    # 一阶bias
    w0 = mx.symbol.var('w0', shape=(1,), init=init_config['w0'],
                       lr_mult=lr_mult_config['w0'], wd_mult=wd_mult_config['w0'])
    
    # wx + b
    y_first_order = mx.symbol.contrib.SparseEmbedding(data=x, weight=w, \
                    input_dim=num_features, output_dim=1)  # None * F * 1
    y_first_order = mx.symbol.broadcast_add(y_first_order.sum(axis=1), w0) # None * F

    # squared terms for subtracting self interactions
    embedding = mx.symbol.contrib.SparseEmbedding(data=x, weight=v, \
                    input_dim=num_features, output_dim=factor_size)  # None * F * K
    # 平方和
    squared_sum_part = mx.symbol.sum(embedding.square(), axis=1)    # None * k
    # 和平方
    sum_squared_part = mx.symbol.square(data=embedding.sum(axis=1)) # None * K
    y_second_order   = 0.5* (sum_squared_part - squared_sum_part)   # None * K

    # putting everything together
    output = mx.symbol.Concat(y_first_order, y_second_order, dim=1)
    model  = output.sum(axis=1, keepdims=True)

    y = mx.symbol.Variable("softmax_label")
    model = mx.symbol.LogisticRegressionOutput(data=model, label=y)`

**Comment by eric-haibin-lin** [comment]
What are you comparing to? What optimizer are you using?

**Comment by charleshm** [comment]
adam, the model didn't converge.  I first wrote fm using tensorflow, it's ok.  but in mxnet, using SparseEmbedding, it didnt converge; using Embedding, it started to converge , but the result was unexpected.

**Comment by eric-haibin-lin** [comment]
I see. Are you using CPU / single GPU / multiple GPUs? 
One thing to notice is that our Adam optimizer by default performs lazy update if it sees sparse weight & sparse grad, which is similar to tf.LazyAdamOptimizer and behaves different from standard Adam optimizer. https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/optimizer.py#L675-L688

There're some effort supporting standard updates with sparse grad, but the one for Adam is not there yet. But it will be available soon  (cc @ZiyueHuang )

> using Embedding, it started to converge , but the result was unexpected.

What do you mean by unexpected result?

**Comment by ZiyueHuang** [comment]
Sure. Standard Adam update is on my list. Will work on it later.

**Comment by charleshm** [comment]
Great !
